File: src/main/java/io/mfedirko/learning/MockLearningRepository.java
package io.mfedirko.learning;

import com.github.rjeschke.txtmark.Processor;
import org.springframework.context.annotation.Profile;
import org.springframework.stereotype.Repository;

import java.time.*;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@Repository
@Profile("mock")
public class MockLearningRepository implements LearningRepository {
    private final List<Lesson> lessons = new ArrayList<>();

    @Override
    public List<Lesson> findLessons(Year date) {
        LocalDateTime startOfYear = date.atMonthDay(MonthDay.of(Month.JANUARY,1)).atStartOfDay();
        return lessons.stream()
                .filter(l -> l.getCreationTimestamp().isAfter(startOfYear))
                .sorted(Comparator.comparing(Lesson::getCreationTimestamp))
                .collect(Collectors.toList());
    }

    @Override
    public Lesson getLesson(long creationTimeMillis) {
        return lessons.stream()
                .filter(l -> l.getCreationTimestampMillis() == creationTimeMillis)
                .findFirst()
                .orElseThrow(() -> new IllegalArgumentException("Lesson not found!"));
    }

    @Override
    public long createLesson(CreateLessonForm lesson) {
        Lesson newLesson = Lesson.builder()
                .parsedDescription(Processor.process(lesson.getDescription()))
                .description(lesson.getDescription())
                .category(lesson.getCategory())
                .title(lesson.getTitle())
                .author("Michael Fedirko")
                .creationTimestamp(LocalDateTime.now())
                .creationTimestampMillis(Instant.now().toEpochMilli())
                .build();
        lessons.add(newLesson);
        return newLesson.getCreationTimestampMillis();
    }

    @Override
    public void updateLesson(UpdateLessonForm lesson, long id) {
        Lesson oldLesson = getLesson(id);
        Lesson newLesson = Lesson.builder()
                .parsedDescription(Processor.process(lesson.getDescription()))
                .description(lesson.getDescription())
                .category(lesson.getCategory())
                .title(lesson.getTitle())
                .author(oldLesson.getAuthor())
                .creationTimestamp(LocalDateTime.now())
                .creationTimestampMillis(Instant.now().toEpochMilli())
                .build();
        deleteLesson(id);
        lessons.add(newLesson);
    }

    @Override
    public void deleteLesson(long creationTimeMillis) {
        lessons.removeIf(l -> l.getCreationTimestampMillis() == creationTimeMillis);
    }
}

File: src/main/kotlin/io/mfedirko/learning/MockLearningRepository.kt
package io.mfedirko.learning

import com.github.rjeschke.txtmark.Processor
import io.mfedirko.common.util.Dates.toUtcStartOfYear
import org.springframework.context.annotation.Profile
import org.springframework.stereotype.Repository
import java.time.*

@Repository
@Profile("mock")
class MockLearningRepository : PaginatedLearningRepository {
    private val lessons: MutableList<Lesson> = ArrayList()

    override fun findLessons(limit: Int, offset: Int): List<Lesson> {
        return lessons
            .sortedWith(compareBy { it.creationTimestamp })
            .drop(offset)
            .take(limit)
            .toList()
    }

    override fun findLessons(year: Year): List<Lesson> {
        val startOfYear = year.toUtcStartOfYear()
        return lessons
            .filter { it.creationTimestamp.isAfter(startOfYear) }
            .sortedWith(compareBy { it.creationTimestamp })
            .toList()
    }

    override fun getLesson(id: Any): Lesson {
        return lessons.first { it.id == id }
    }

    override fun createLesson(lesson: CreateLessonForm): Any {
        val newLesson: Lesson = Lesson().apply {
            description = lesson.description!!
            parsedDescription = Processor.process(lesson.description)
            category = lesson.category!!
            title = lesson.title!!
            author = "Michael Fedirko"
            creationTimestamp = LocalDateTime.now()
            id = Instant.now().toEpochMilli()
        }
        lessons.add(newLesson)
        return newLesson.id
    }

    override fun updateLesson(lesson: UpdateLessonForm, id: Any) {
        val oldLesson = getLesson(id)
        val newLesson: Lesson = Lesson().apply {
            parsedDescription = Processor.process(lesson.description)
            description = lesson.description!!
            category = lesson.category!!
            title = lesson.title!!
            author = oldLesson.author
            creationTimestamp = LocalDateTime.now()
            this.id = Instant.now().toEpochMilli()
        }
        deleteLesson(id)
        lessons.add(newLesson)
    }

    override fun deleteLesson(id: Any) {
        lessons.removeIf { it.id == id }
    }
}